src/v/storage/spill_key_index.cc

```diff
@@ -88,6 +88,35 @@ ss::future<> spill_key_index::index(
     });
 }
 
+ss::future<>
+spill_key_index::spill_some(size_t entry_size, size_t min_index_size) {
+    auto stop_cond = [this, entry_size, min_index_size] {
+        size_t total_mem = idx_mem_usage() + _keys_mem_usage + entry_size;
+
+        // Instance-local capacity check
+        bool local_ok = total_mem < _max_mem;
+
+        // Shard-wide capacity check
+        bool global_ok = _resources.compaction_index_bytes_available()
+                         || total_mem < min_index_size;
+
+        // Stop condition: none of our size thresholds must be violated
+        return _midx.empty() || (local_ok && global_ok);
+    };
+
+    while (!stop_cond()) {
+        /**
+         * Evict first entry, we use hash function that guarante good
+         * randomness so evicting first entry is actually evicting a
+         * pseudo random elemnent
```

Comment:
as Travis pointed out in the previous PR, this isn't necessarily true. It's just copied here verbatim in the first refactor commit.